mixnet/client/src/sender.rs

```diff
@@ -133,21 +133,58 @@ where
             return None;
         }
 
+        self.remaining_delays -= 1;
+
+        // corner case: 0 average delay
+        if self.avg_delay == 0 {
+            return Some(0);
+        }
+
         if self.remaining_delays == 1 {
-            self.remaining_delays -= 1;
             return Some(self.remaining_time);
         }
 
+        // if we have no remaining time, then just return 0
+        if self.remaining_time == 0 {
+            return Some(0);
+        }
+
         // Calculate bounds to avoid extreme values
-        let lower_bound = (self.avg_delay as f64 * 0.5) as u64;
         let upper_bound = (self.avg_delay as f64 * 1.5)
-            .min((self.remaining_time - self.remaining_delays + 1) as f64)
-            as u64;
+            // guarantee that we don't exceed the remaining time and promise the delay we return is
+            // at least 1ms.
+            .min((self.remaining_time as f64 - self.remaining_delays as f64) + 1.0);
+        let lower_bound = (self.avg_delay as f64 * 0.5).min(upper_bound);
 
-        let delay = self.rng.gen_range(lower_bound, upper_bound);
-        self.remaining_time -= delay;
-        self.remaining_delays -= 1;
+        let delay = self.rng.gen_range(lower_bound, upper_bound + 1.0).floor() as u64;
```

Comment:
not a fan of +1, as you need to do some calculations to show that it's safe to use and it's easy to mess it up with future refactors. The next api would be better, but since we have to use an old version of rand, I suggest we use https://docs.rs/rand/0.7.3/rand/distributions/struct.Uniform.html#method.new_inclusive

Reply:
Good suggestion! Thanks!

Reply:
Good suggestion! Thanks!